dredis/keyspace.py

```diff
@@ -157,41 +160,49 @@ def _delete_db_set(self, key):
         # there are two sets of db keys for sets:
         # * set
         # * set members
+        #
+        # currently the `set` key is immediately deleted and the other keys
+        # will be collected by gc.KeyGarbageCollector()
+        key_id, _ = self._get_set_key_id_and_length(key)
         with self._db.write_batch() as batch:
             batch.delete(KEY_CODEC.encode_set(key))
-            for db_key, _ in self._get_db_iterator(KEY_CODEC.get_min_set_member(key)):
-                batch.delete(db_key)
+            batch.put(KEY_CODEC.encode_deleted_set(key_id), bytes(''))
 
     def _delete_db_hash(self, key):
         # there are two sets of db keys for hashes:
         # * hash
         # * hash fields
+        #
+        # currently the `hash` key is immediately deleted and the other keys
+        # will be collected by gc.KeyGarbageCollector()
+        key_id, _ = self._get_hash_key_id_and_length(key)
         with self._db.write_batch() as batch:
             batch.delete(KEY_CODEC.encode_hash(key))
-            for db_key, _ in self._get_db_iterator(KEY_CODEC.get_min_hash_field(key)):
-                batch.delete(db_key)
+            batch.put(KEY_CODEC.encode_deleted_hash(key_id), bytes(''))
 
     def _delete_db_zset(self, key):
         # there are three sets of db keys for zsets:
         # * zset
         # * zset scores
         # * zset values
+        #
+        # currently the `zset` key is immediately deleted and the other keys
+        # will be collected by gc.KeyGarbageCollector()
+        key_id, _ = self._get_zset_key_id_and_length(key)
         with self._db.write_batch() as batch:
             batch.delete(KEY_CODEC.encode_zset(key))
-            for db_key, _ in self._get_db_iterator(KEY_CODEC.get_min_zset_score(key)):
-                batch.delete(db_key)
-            for db_key, _ in self._get_db_iterator(KEY_CODEC.get_min_zset_value(key)):
-                batch.delete(db_key)
+            batch.put(KEY_CODEC.encode_deleted_zset_score(key_id), bytes(''))
+            batch.put(KEY_CODEC.encode_deleted_zset_value(key_id), bytes(''))
```

Comment:
What are these lines doing? Is it these 2 puts in your example, so marking the entire zset + unique ID to be garbage collected?

```
[del] 6_z = uniqueID_2
[put] x_7_uniqueID = ''
[put] x_8_uniqueID = ''
```

Reply:
Exactly, @jimjshields. The GC is based on prefixes, so it will look for all keys related to scores and values separately (they have different prefixes).